widgetastic_manageiq/__init__.py

```diff
@@ -2174,3 +2179,40 @@ class ProviderEntity(BaseEntity):
     quad_entity = ProviderQuadIconEntity
     list_entity = ProviderListEntity
     tile_entity = ProviderTileIconEntity
+
+
+class DynamicTable(VanillaTable):
+    """Extend the widget.Table class to implement row_add for dynamic tables with an 'Actions'
+    column.
+    In these tables, the top or bottom row can be clicked to add a new row, and when it is
+    clicked the row is replaced (top or bottom) with a row containing fillable widgets.
+
+    Args:
+        action_row: index of the action row, generally 0 or -1, defaults to 0
+
+        See Widgetastic.widget.Table for more arguments
+    """
+
+    def __init__(self, *args, **kwargs):
+        self.action_row = kwargs.pop('action_row', 0)  # pull this off and pass the rest up
+        super(DynamicTable, self).__init__(*args, **kwargs)
+
+    def row_add(self):
+        """Use the action-cell column widget to add a row
+
+        Clicks on the row directly, not the action button
+        """
+        try:
+            self[self.action_row].click()
+            return self.action_row
```

Comment:
So, yes, getitem handles it, but only for itself, so if action_row is negative, getitem will correctly handle it, but then you return action_row again, which is always pointing to the "add row" when negative. So my concern is that when new row is added, and the number is negative, it will return the wrong row unless the row turns itself into the editing mode, and by returning the negative, the machinery that uses the result of this method's call will not have a stable anchor point since if anything changes in the table, it will suddenly operate on a different row. So I would pre-process the action_row with the _process_negative_index here as well to have a positive index :)